delft3dworker/models.py

```diff
@@ -479,11 +490,160 @@ class Container(models.Model):
+        different from the desired_state, this Container will act: it will
+        start a task to get both states matched.
+
+        At the end the Container will request a log update.
+        """
+
+        self._update_state_and_save(snapshot)
+
+        if self.task_uuid == '':
+
+            self._fix_state_mismatch()
+
+            self._update_log()
+
+    def _update_state_and_save(self, snapshot):
+        """
+        Var snapshot can be either dictionary or None.
+        If None: docker container does not exist
+        If dictionary: snapshot['Status'] is a string describing status
+        """
+
+        if snapshot is None:
+            self.docker_state = 'non-existent'
+
+        elif isinstance(snapshot, dict) and ('Status' in snapshot):
+
+            if snapshot['Status'].startswith('Up'):
+                self.docker_state = 'running'
+
+            elif snapshot['Status'].startswith('Created'):
+                self.docker_state = 'created'
+
+            elif snapshot['Status'].startswith('Exited'):
+                self.docker_state = 'exited'
+
+            else:
+                self.docker_state = 'unknown'
+
+        else:
+            self.docker_state = 'unknown'
+
+        self.save()
+
+    def _fix_state_mismatch(self):
+        """
+        If the docker_state differs from the desired_state, and Container is
+        not waiting for a task result, a task should be executed to fix this
+        mismatch.
+        """
+
+        # return if container still has an active task
+        if self.task_uuid is not None:
+            return
+
+        # return if the states match
+        if self.desired_state == self.docker_state:
+            return
+
+        # appearantly there is something to do, so let's act:
+
+        if self.desired_state == 'created':
+            self._create_container()
+
+        if self.desired_state == 'running':
+            self._start_container()
+
+        if self.desired_state == 'exited':
+            self._stop_container()
+
+        if self.desired_state == 'non-existent':
+            self._delete_container()
+
+    def _create_container(self):
+        if self.docker_state != 'non-existent':
+            return  # container is already created
+
+        # delay the create task with the appropriate image name
+        image_dict = {
+            'delft3d': settings.DELFT3D_IMAGE_NAME,
+            'export': settings.EXPORT_IMAGE_NAME,
+            'postprocess': settings.POSTPROCESS_IMAGE_NAME,
+            'preprocess': settings.PREPROCESS_IMAGE_NAME,
+            'process': settings.PROCESS_IMAGE_NAME
+        }
+        result = do_docker_create.delay(
+            image_dict[self.container_type]
+        )
+
+        self.task_uuid = result.id
+        self.save()
+
+    def _start_container(self):
+        if self.docker_state != 'created' and self.docker_state != 'exited':
+            return  # container is not ready for start
```

Comment:
If a container is not ready, there is faulty logic somewhere, please use log.warning